### PR TITLE
wayland: ensure external loop is notified with pump_events

### LINKED ISSUE #4183: Cannot integrate winit wayland backend into existing event loop
### Description

To integrate, the event loop needs not only the wl_display but also the wl_event_queue used by winit so that it can call wl_display_prepare_read_queue before going to sleep. This requirement is described, for example, here: https://gitlab.freedesktop.org/wayland/wayland/-/blob/6137c8c21383d7ebb92cf3331be56ca0aecccd5d/src/wayland-client.c?page=2#L1863

But winit doesn't expose its wl_event_queue. 

### Debugging output

```shell

```

### Window isn't shown unless you draw

- [x] I understand that windows aren't shown on Wayland unless I draw and present to them.

### Winit version

0.30.9

### PR DESCRIPTION
**Pull request by kchibisov**
Fixes #4183.

### CODE REVIEW

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 1]
```diff
@@ -68,6 +71,56 @@ pub struct EventLoop {
...
+fn spawn_wakeup_thread(
+    connection: Connection,
+    awakener: calloop::ping::Ping,
+) -> Option<PumpEventNotifier> {
+    // Start from the waiting state.
+    let control = Arc::new((Mutex::new(true), Condvar::new()));
+    let control_thread = Arc::clone(&control);
+
+    let handle = std::thread::Builder::new()
+        .name(String::from("wayland pump_events wake-up notify"))
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:95]: The first 15 character should contain enough information to identify the thread since everything after that won't be visible in non-rust APIs.

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 2]
```diff
@@ -68,6 +71,56 @@ pub struct EventLoop {
...
+        .spawn(move || {
+            let (lock, cvar) = &*control_thread;
+            loop {
+                let mut wait = lock.lock().unwrap();
+                while *wait {
+                    wait = cvar.wait(wait).unwrap();
+                }
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:102]:
```suggestion
                {
                    let mut wait = lock.lock().unwrap();
                    while *wait {
                        wait = cvar.wait(wait).unwrap();
                    }
                    *wait = true;
                }
```

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 3]
```diff
@@ -68,6 +71,56 @@ pub struct EventLoop {
...
+
+                if let Some(read_guard) = connection.prepare_read() {
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:104]:
```suggestion
                while let Some(read_guard) = connection.prepare_read() {
```

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 4]
```diff
@@ -191,6 +238,12 @@ impl EventLoop {
         timeout: Option<Duration>,
         mut app: A,
     ) -> PumpStatus {
+        if let Some(pump_event_notifier) = self.pump_event_notifier.as_ref() {
+            // Notify that we should start waiting, since we're in winit.
+            *pump_event_notifier.control.0.lock().unwrap() = true;
+            pump_event_notifier.control.1.notify_one();
+        }
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:245]:
```suggestion
```

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 5]
```diff
@@ -209,6 +262,37 @@ impl EventLoop {
...
+
+                if self.pump_event_notifier.is_some() {
+                    let wayland_dispatcher = self.wayland_dispatcher.clone();
+                    self.event_loop
+                        .handle()
+                        .insert_source(
+                            event_loop_awakener_source,
+                            move |_, _, winit_state: &mut WinitState| {
+                                let mut source = wayland_dispatcher.as_source_mut();
+                                let _ = winit_state.dispatch_pending(source.queue());
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:284]: This seems unnecessary.

**Review by mahkoh — CHANGES_REQUESTED** [review]
Haven't tested it.

**Review by mahkoh — DISMISSED** [review]
Haven't tested it.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 5 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:284]: it's necessary. The dispatch isn't done otherwise IIRC, but I can test one more time though.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 4 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:245]: yeah, that is redundant, though, I left for consistency.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 3 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: I think that is redundant? Like you'll notify winit one way or another, no need to get in `while`.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: You should only trigger a notification if `prepare_read` returns `None` since otherwise there are no messages to dispatch.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 5 above)
Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:284]: Doesn't `pump_app_events` always dispatch events?

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 4 above)
Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:245]: You shouldn't wake up the thread before having dispatched all messages since it will just immediately trigger a notification again.

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 6]
```diff
@@ -68,6 +71,56 @@ pub struct EventLoop {
...
+        .spawn(move || {
+            let (lock, cvar) = &*control_thread;
+            loop {
+                let mut wait = lock.lock().unwrap();
+                while *wait {
+                    wait = cvar.wait(wait).unwrap();
+                }
+
+                if let Some(read_guard) = connection.prepare_read() {
+                    let _ = connection.flush();
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:105]: This needs to be a loop with polling until flush returns Ok.

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 7]
```diff
@@ -68,6 +71,56 @@ pub struct EventLoop {
...
+
+#[derive(Debug)]
+struct PumpEventNotifier {
+    /// Whether we're in winit or not.
+    control: Arc<(Mutex<bool>, Condvar)>,
+    /// Thread handle.
+    _handle: JoinHandle<()>,
+}
+
+fn spawn_wakeup_thread(
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:86]: The thread needs to be shut down at some point.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 5 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:284]: Yeah, it's required. I think it's a bug though, should be fixed on `calloop` end.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 6 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:105]: This should likely mention that, but I've never seen a wayland loop that `flush` in a loop until it works. I'd also assume that it may fail in a non-recoverable way.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 3 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: Also, from my testing `prepare_read` never returned `-1` since the queue is always empty, since queue ends up having zero events, because winit's main loop has a different queue.

If I had an option to use that queue `poll` won't be needed either. So the way you've suggested straight won't work. I may retry only on `flush` failure, but the way it's suggested doesn't really work for us.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: >Also, from my testing prepare_read never returned -1 since the queue is always empty, since queue ends up having zero events, because winit's main loop has a different queue.

You have to use winit's queue here.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 4 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:245]: it would have been blocked in `read_events`, otherwise we'd had events to read and thus wake-up would have been needed anyway. But yeah should be fine.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 3 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: I'm well aware about it, I can not though because of API, but this also _works_ and one can check the `return` value of the `read_events` if that's your concern, since it still tells you that you've read something.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: Then we have no way to fix this, unless `wayland-rs` is patched, I simply has no API to do anything about this.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: Actually, if we always call `read_events` and `read` with a poll in a loop, there will be a way to fix it, since we do blocking API polling, thus will be fine.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:104]: >Then we have no way to fix this, unless wayland-rs is patched

It seems that way. I'm currently working on a new, safe libwayland wrapper that will avoid all of this.

>Actually, if we always call read_events and read with a poll in a loop, there will be a way to fix it

I'll believe it when I see it.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: Yeah, it'll work, but end up with over wake-ups. I think I'd just go and patch `wayland-rs`, not a big issue for me and I know how to easily do it.

>It seems that way. I'm currently working on a new, safe libwayland wrapper that will avoid all of this.

The question is whether will it go, I doubt that it'll be shared in smithay. Though, if your thing will have API that is easier to use, I'll switch over tbf.

**Comment by kchibisov** [comment]
I think now it should work, though, I'll need a patch for `wayland-sys`.

**Comment by kchibisov** [comment]
@mahkoh did you have a chance to look into it, I think it should work correctly, if so I'll try to propagate fix through out and make a patch release.

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 8]
```diff
@@ -697,3 +719,76 @@ impl rwh_06::HasDisplayHandle for OwnedDisplayHandle {
...
+                            PumpEventNotifierAction::Pause => (),
+                        }
+
+                        wait = cvar.wait(wait).unwrap();
+                    }
+                    // Wake-up the main loop and put this one back to sleep.
+                    *wait = PumpEventNotifierAction::Pause;
+                    drop(wait);
+
+                    while let Some(read_guard) = connection.prepare_read() {
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:767]: Does this lock the queue used by winit?

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:767]: Yeah, it all uses the same queue, thus we fall into block.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:767]: How does this function work if an application creates multiple queues? Does it lock all of them?

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:767]: wayland-rs uses the single queue through out and dispatches into internal stuff from its internal queue. The function I've added just calls `read_events` and thus will pull into `wayland-rs` queue internal to its `-sys` backend. The later dispatches dispatch from this queue into other queues.

So if you have multiple queues there shouldn't be an issue from what I understood, which have a separate thread that deals only with `wayland-rs` queue. If you create something else from ffi with `wayland-rs` you'll end up with different queue. The `queue` in wayland-rs I'm talking about, is a queue internal to wayland-rs and from which it dispatches to its own Rust based `Queue`.

https://github.com/Smithay/wayland-rs/pull/802/files

Like it was always the case that `wayland-rs` used a single _internal_ queue(the one backed by `wl_event_queue`) from which it dispatched to `Queue` stuff, but it always dispatched things.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:767]: Alright.

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 9]
```diff
@@ -697,3 +719,76 @@ impl rwh_06::HasDisplayHandle for OwnedDisplayHandle {
...
+    }
+}
+
+impl PumpEventNotifier {
+    fn spawn(connection: Connection, awakener: Ping) -> Self {
+        // Start from the waiting state.
+        let control = Arc::new((Mutex::new(PumpEventNotifierAction::Pause), Condvar::new()));
+        let control_thread = Arc::clone(&control);
+
+        let _ = connection.prepare_read().unwrap();
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:747]: What is the purpose of this?

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 10]
```diff
@@ -697,3 +719,76 @@ impl rwh_06::HasDisplayHandle for OwnedDisplayHandle {
...
+                        wait = cvar.wait(wait).unwrap();
+                    }
+                    // Wake-up the main loop and put this one back to sleep.
+                    *wait = PumpEventNotifierAction::Pause;
+                    drop(wait);
+
+                    while let Some(read_guard) = connection.prepare_read() {
+                        let _ = connection.flush();
+                        let poll_fd = PollFd::from_borrowed_fd(connection.as_fd(), PollFlags::IN);
+                        let _ = rustix::event::poll(&mut [poll_fd], -1);
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:770]: This will cause the join to deadlock.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 9 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:747]: I think I just left it for some reason and completely forgot, sorry.

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 10 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: Added pipe to wake-up the thread just in case.

File: src/platform_impl/linux/wayland/event_loop/mod.rs [hunk 11]
```diff
@@ -697,3 +721,84 @@ impl rwh_06::HasDisplayHandle for OwnedDisplayHandle {
...
+                'outer: loop {
+                    let mut wait = lock.lock().unwrap();
+                    while *wait == PumpEventNotifierAction::Pause {
+                        wait = cvar.wait(wait).unwrap();
+                    }
+                    // Wake-up the main loop and put this one back to sleep.
+                    *wait = PumpEventNotifierAction::Pause;
+                    drop(wait);
+
+                    while let Some(read_guard) = connection.prepare_read() {
```
Comment (mahkoh) [inline src/platform_impl/linux/wayland/event_loop/mod.rs:773]: Does this also ensure that all wayland-client event queues are empty?

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: it just calls the `prepare_read_queue` on the `wl_event_queue` used by `wayland-client`, so the logic is exactly as you'd use the C calls directly.

it just, the issue was with `read` before since it had `dispatch` unconditionally.

The logic right now is that if you have something in the queue (you read it before with `read_events`) then it'll go to notifying the loop and going back to sleep.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: But generally, if the queue was not empty, we'll just notify the event loop without doing anything extra. The `None` is returned when you have something in the `queue` (as you'd get `-1`).

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: >it just calls the prepare_read_queue on the wl_event_queue used by wayland-client, so the logic is exactly as you'd use the C calls directly.

In C there is only one queue layer. It wayland-client there are two queue layers and you have to make sure that both are empty before going to sleep.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: Do you mean the internal `libwayland-client` one? Or do you mean the `wl_event_queue` in `wayland-rs` and whatever other abstraction you have?

In general, once we re-enter winit we dispatch everything we're read on the pump events thread, there could be still not read events though (on socket side), but as for how this thread will go, I don't think it's an issue. Things can not move into `wayland-rs` rust side queue until you call the `dispatch_inner_queue` which only the main thread with `pump_events` calls.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: The queue this thread touches is only the `wl_event_queue` we've created with the `C` API, it doesn't interact with any other queue.

The `pump_events` thread, is the one pulling from `wl_event_queue` into the `Queue`, but it dispatches everything straight away.

The only difference on this thread compared to C, is that we don't use default queue, but I think it shouldn't matter?

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: Though, I'm not sure how the wiring that `wayland-rs` does interacts when other thread pulls everything, I'd assume that it'll end up in `wl_event_queue`, and only once we call the `dispatch_inner_queue` it'll advance further.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: >The pump_events thread (where EventLoop is), is the one pulling from wl_event_queue into the Queue, but it dispatches everything straight away.

This still relies on there only being one thread and one Queue. Otherwise any other thread could call dispatch_inner which would read events into any other Queue. It is fragile in a way that the C API is not.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: All `Queue` share the same `wl_event_queue`, if you pass anything `wayland-rs` to ffi and create from it, it'll create a new `wl_event_queue`, so they are all isolated.

So effectively, if you use `wayland-rs` with winit and create from its `ffi` object, you create a new `wl_event_queue` and so can not mess up the queue in winit. The only issue could happen _if winit will share the rust objects directly, which will likely never happen.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: >The only issue could happen _if winit will share the rust objects directly, which will likely never happen.

It could also happen if there is more than one Queue. I.e. the wayland-client API cannot be used if an application uses multiple Queues.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: Are you talking about general issue or the issue with the concret implementation right now, since right now I don't see how there could be an issue given that winit is isolated from any other code.

Like any user will have different C queue, even if you test with `pump_events` example it also has 2 queues, the one in `softbuffer` and the one in `winit` and there's no issue with such setup.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: >Are you talking about general issue or the issue with the concret implementation right now, since right now I don't see how there could be an issue given that winit is isolated from any other code.

I don't know what winit does. But anything that relies on there never being more than 1 event queue seems questionable.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: But right now we already have more than one queue in tests, but queues isolated from each other... The issue is only if I within winit's code start adding extra stuff within itself and I somehow do the polling the way it was originally done on this PR.

I don't see how adding extra C queue outside of winit will mess up winit. It'll basically read into `wl_event_queue` (C object) winit has and then we'll start getting `None` in `prepare_read` (the way it worked with blocking poll).

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: I'm not talking about wl_event_queue. I'm talking about wayland-client queues.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: But they are backed by `wl_event_queue`, and once you create new `Connection` from `ptr` it's backed by new `wl_event_queue`. It's true that multiple `Queue` refer to the same `wl_event_queue`, but they are encapsulated by `Connection` object. So you can not mess with winit `Queue` outside of winit, since you are all separated by `Connection`.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: I'm aware.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: yeah, I'm then not sure how winit state could be messed up by some other thread, since only winit can call into `dispatch_inner_queue` to dispatch its `wl_event_queue`.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: I'm saying that it is fragile in a way that the C API is not. The C API supports any number of threads using any number of queues. The wayland-client API does not. If anyone at any point decides to add more queues to to winit, the polling will silently break.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs:773]: Yeah, that part I understand, so we agree on that point. I'm just saying that in the current state of things there's no _issue_. It's true that someone will mess up with queues on my behalf and not notice that example broke (if you add a queue the thing that will broke is the `pump_events` example straight away).

File: src/platform_impl/linux/wayland/event_loop/mod.rs (hunk 3 above)
Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: https://docs.rs/wl-client in particular https://docs.rs/wl-client/latest/wl_client/struct.QueueWatcher.html would be used to solve this issue.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: >Also, is there really no way to pass just &mut self during dispatch?

No, wayland event handlers are reentrant. They cannot provide a mutable reference unless you make compromises elsewhere.

>I guess the reason is the async so you're forced to do &self?

It's unrelated.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: Correct me if I'm wrong, but if you read the input sequentially, you'll also callback sequentially, thus having `&mut` is _safe_? But it really depends what the `&self` is here.

I just really don't like that everything must be pretty much a `borrow` and that `State` is not entirely centralized. Like the modal `wayland-rs` has is not bad one in particular, I'm not sure why we can not have e.g. `&mut State` passed around callbacks, and e.g. not use `udata` or if use `udata` use it only for non-mut access and imply that it's `Send`.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: >Correct me if I'm wrong, but if you read the input sequentially, you'll also callback sequentially, thus having &mut is safe? But it really depends what the &self is here.

You can dispatch the queue from within an event handler which might invoke the event handler again.

>Like the modal wayland-rs has is not bad one in particular, I'm not sure why we can not have e.g. &mut State

The wayland-client model means that you cannot easily have per-proxy data. You can have some kind of index into a hash map, but then you have to look up the per-proxy data in the hash map every time you get an event. And if you then need to pass the `&mut State` to some other function that might modify that map, you must first release access to the per-proxy data and re-acquire it afterwards etc.

In wl-client you can easily have per-proxy data but you only get a `&self` so some kind of interior mutability has to be used.

The wl-client model is much less invasive in my experience plus it avoids having to have two queue layers (which also messes up WAYLAND_DEBUG output) plus it avoids the per-event allocations that wayland-client performs to be able to store events in the second queue layer indefinitely when libwayland might already have deallocated the original message.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: >You can dispatch the queue from within an event handler which might invoke the event handler again.

Hm, but do one really need that? I'm not sure I've even seen C code that does so. And also, if you require `&mut State` to be passed to `dispatch` won't it just work? Since you just passed it e.g. through?

Though, I guess it would be hard to restrict in terms of rust type system.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: I'd personally generally prefer an API without interior mutability, since debugging interior mutability can be done only afterwards in runtime in case of some bug happens, and generally, I'd like to have less of it not more.

Having state just be `RefCell` because of code (I'd personally) never write is a bit unfortunate.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: >Hm, but do one really need that?

It happens all the time. E.g. in wl_registry::global you might bind a global and then perform a roundtrip to make sure that you've seen all of the initial events of the new object.

>And also, if you require &mut State to be passed to dispatch won't it just work?

Then you cannot easily have per-proxy data as described above.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: I guess we just have different samples here. Most code I've seen does it in a way that it reads globals, binds, and then goes into `loop {}`.

>Then you cannot easily have per-proxy data as described above.

per-proxy data being anything other than readonly set at the time of creation is unmanageable anyway. Having `&mut State` is a safer and saner way to write it anyway. Like the only use case for proxy data I have is e.g. set a `Key` into `HashMap` or some other minor data there that never changes or changes with some minor interior mutability.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: yeah, I just everything into one giant state, since once you need to access something from other objects any proxy data is getting ugly pretty fast.

but it's just a preference, I write client that way, some other people write clients other way. but having `&mut` state with all the data need and leave proxy data readonly pretty much relatively easy to manage.

Comment (kchibisov) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: I've looked more into the source code of the `wl-client` and I think even `&mut` doesn't really help me since of the design to have unique handler for each type of event basically.

I don't see a big benefit in changing at the current state, since the API is wastly different. Though, in the future once we've done with trait based design, one can write a `wl-client` based backend.

Since at the current state of things one can not simply adopt the design, it needs a full rewrite pretty much, for which I don't have time to just compare what performs better. Though, the design with separate callback and event handlers looks like an old wayland-rs, which was really bad in terms of maintainability and dealing with interior mutability, but it was long ago so my memory is loose here.

I really don't like dealing with interior mutability, so if there was an option along side present once to allow `&mut` somehow through out and so I can pass the `&mut State` then I'd be more inlined to try to port myself and see how it performs. My usage is basically everything wayland related lives on the same thread and that I never dispatch from within handling of the event.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: >I really don't like dealing with interior mutability, so if there was an option along side present once to allow &mut somehow through out and so I can pass the &mut State

It would be trivial to add an option to wl-client-builder that creates an adapter that wraps your event handler in a RefCell. It would panic if you do have a recursive dispatch and would not be any different from you using such a RefCell yourself, so I don't see too much value in it.

Comment (mahkoh) [reply src/platform_impl/linux/wayland/event_loop/mod.rs]: It probably wouldn't be too hard to add a `&mut Data` parameter.
